### PR TITLE
Fixes #33470 - Adding support for CentOS Stream

### PR DESCRIPTION
**Pull request by domitea**
The facts from Stream has not yet supported in Foreman. This PR adds correct recognition for all parsers.

### CODE REVIEW

File: test/unit/foreman_salt/salt_fact_parser_test.rb [hunk 1]
```diff
@@ -23,6 +23,17 @@ def setup
...
+      facts = HashWithIndifferentAccess.new(read_json_fixture("facts/salt_centos_stream.json"))
+      parser = ForemanSalt::FactParser.new(facts)
+      os = parser.operatingsystem
+
+      assert os.present?
+      assert_equal 'CentOSStream', os.name
+      assert_equal '8', os.major
+      assert_empty os.minor
+    end
+
```
Comment (lzap) [inline test/unit/foreman_salt/salt_fact_parser_test.rb:36]: Please add a test case for "CentOS 8 LInux" too to ensure this patch does not add regression for that.

You can find example JSON here: https://github.com/theforeman/foreman/pull/8456/files

**Comment by lzap** [comment]
Does this replace https://github.com/theforeman/foreman/pull/8456 ?

**Comment by domitea** [comment]
Well, the #8456 is aimed only for puppet parser. This PR affects all parsers.

**Comment by lzap** [comment]
Please take Katello OS creation into cosideration. Ideally, we should create the same name as Katello, which I believe (test this) creates "CentOS_Stream" when you sync a CentOS Stream repo. These need to match. One possibility to solve this is to patch Katello to create this particular OS correctly. Or ideally, Foreman core would expose a class/method that would allow mapping of kickstart OS names to Foreman names.

**Comment by ares** [comment]
@domitea this may give you a pointer, Katello [creates](https://github.com/Katello/katello/blob/master/app/models/katello/concerns/redhat_extensions.rb#L7-L50) the OS here, when it [downloads](https://github.com/Katello/katello/blob/master/app/services/katello/pxe_files_downloader.rb#L11). I think the hint with the underscore is correct, we should use the same notation.

**Comment by ares** [comment]
And just confirmed on the demo HW, Katello creates "CentOS_Stream 8" after the BaseOS repo sync

**Comment by domitea** [comment]
The tests for Katello and Foreman are green so I think it's ready to merge.

**Comment by ShimShtein** [comment]
@lzap ACK from me, if you have no further comments, we can merge it.

Awesome work @domitea! It's really good to see that you have taken all parsers into account

**Review by ezr-ondrej — APPROVED** [review]
Thanks @domitea for this :) Thanks @ShimShtein, @ares and @lzap for reviews :+1: 
Sorry @lzap for not waiting for you to merge :)

**Comment by lzap** [comment]
> Sorry @lzap for not waiting for you to merge :)

Dont apologize, just be ready to fix things if they beak and I am good :) Thanks for pushing this forward.

**Comment by ekohl** [comment]
Personally I still think this is the wrong direction. CentOS Linux 8 is EOL already. It will disappear from mirrors in 3 weeks. IMHO we should revert this and go back to the way it was.

It certainly should not be cherry picked into a stable release.

**Comment by ezr-ondrej** [comment]
Unfortunatelly I didn't pick this comment before todays call.
I agree there is not a good way. There might be something better then this, but this unified all reporters to report the same thing, which is a big deal IMHO. We might decide that this thing is not the correct one, but it should definitelly be unified. I do not feel strongly about this and I would be fine if the same thing is `CentOS` for stream :shrug:

**Comment by lzap** [comment]
I am not sure, CentOS 8 will not "disappear", it will stay there and eventually it will be moved to vault mirror just like other version.

Ewoud, you haven't articulated a single reason in this PR why this should not to be merged, I am not sure if you are referring to some comments in my original PR but from this perspective we are clear to backport. Not doing anything is worse than anything else, people were coming to us complaining. Unfortunately, we were so slow that this was only merged after 8.5 was released.

**Comment by ekohl** [comment]
> I am not sure, CentOS 8 will not "disappear", it will stay there and eventually it will be moved to vault mirror just like other version.

It will disappear from the mirrors at the end of the month and be moved to the vault. The vault should not be used. We can consider those systems End of Life and unsupported. By the time we'll release the next Foreman version, we're way past that point.

Also note that https://centos.org/centos-linux-eol/ states:
> Additionally, with this deadline falling during a time when many of our team, as well as many of our users, will be out of the office, we intend to bend the usual EOL process by keeping this content available until January 31st.
> 
>At that time, or in the event of a serious security bug in this time window (Defined as anything with a VCSS v3 score of 9 or greater), this content will be removed from our mirrors, and moved to vault.centos.org where it will be archived permanently, since we will not be able to provide updates to the content after the EOL date.

So if a critical security bug shows up tomorrow, it's gone to the vault then.

> Ewoud, you haven't articulated a single reason in this PR why this should not to be merged, I am not sure if you are referring to some comments in my original PR but from this perspective we are clear to backport. Not doing anything is worse than anything else, people were coming to us complaining. Unfortunately, we were so slow that this was only merged after 8.5 was released.

I voiced it clearly in https://github.com/theforeman/foreman/pull/8456 and missed this PR. However, I'll state some reasons:

If we really want to identify CentOS Stream, we should list it in the templates (partition tables, kickstarts) because this is used to associate operating systems to it:
https://github.com/theforeman/foreman/blob/9690f3aef6166fdd41979a167cfe93dd5cc40eab/app/services/foreman/render_templates_from_folder.rb#L65-L66

Since it is now listed as a separate OS, to remain compatible, you should really list CentOS_Stream in those templates too. So this is in fact a regression.

Another point to consider: today there is:

* CentOS Linux 7
* CentOS Stream 8
* CentOS Stream 9

Some users may have CentOS Linux 8 installed. These are End of Life, which means effectively the use case should pretty much be to help users migrate to CentOS Stream 8 or another distribution. Do we really want to do that? And does this do so effectively?

As suggested in https://github.com/theforeman/foreman/pull/8456#pullrequestreview-640936164, I still think we should wait it out and continue using just `CentOS`.

I also dislike the underscore in the name. I hope that doesn't show up in the UI.

That said, I do appreciate that this at least aligns the fact parsers.

**Comment by lzap** [comment]
Your whole argumentation is based on the fact that the moment CentOS 8 is EOL everybody stops using that. I think that's not the case, people have ton of CentOS 6/7/8 systems and they will often continue using unsupported versions because it is too expensive for them to migrate. What happens when CentOS 8 Stream makes some change that will let's say make provisioning incompatible with the traditional CentOS releases? Having these are two distinct Foreman OS entries is a good solution.

This allows us to make a template statement like "if OS is CentOS and version major is 8 then something" or "if OS is CentOS_Stream and major 8 do something else". Note there is no minor version anymore for us to decide this if we had OS name as you suggested just `CentOS`.

> Since it is now listed as a separate OS, to remain compatible, you should really list CentOS_Stream in those templates too. So this is in fact a regression.

https://github.com/theforeman/foreman/pull/9049

> As suggested in [#8456 (review)](https://github.com/theforeman/foreman/pull/8456#pullrequestreview-640936164), I still think we should wait it out and continue using just `CentOS`.

It's not like I really like this version and really hate your version, we just needed to do SOMETHING, we already delivered this late. If you want to go ahead and rename everything before we even start using CentOS_Stream go ahead. I just don't feel this is really important.

> I also dislike the underscore in the name. I hope that doesn't show up in the UI.

Well, I did not like it either and that was the reason why `CentOSStream` was in my original proposal.

**Comment by ekohl** [comment]
I suspect this broke our nightly pipelines: https://community.theforeman.org/t/katello-nightly-rpm-pipeline-1220-failed/27082/2

As for your response.

> Your whole argumentation is based on the fact that the moment CentOS 8 is EOL everybody stops using that.

Correct. Next week it'll disappear from the mirrors. In fact, the whole polkit root exploit that is currently going around will not be patched in CentOS Linux 8. I don't think we need to take care of an OS with an active root exploit.

> What happens when CentOS 8 Stream makes some change that will let's say make provisioning incompatible with the traditional CentOS releases?

Then users can either overwrite the templates but they really should not provision CentOS Linux 8 anymore. I still maintain that by now it's safe to assume that there is only CentOS Stream 8.

Users who want a real rebuild of RHEL will switch over to Rocky, AlmaLinux or OracleLinux. They may use tooling such as https://github.com/almalinux/almalinux-deploy to convert their CentOS Linux 8 in place.

**Comment by ares** [comment]
Thanks for the link to the nightly breakage. For the later response I'd like to ask everyone involved to stop the discussion in the closed PR. If this is something you two want to clarify among yourselves, there are better communication channels for it. If you want to have this as a public discussion, please move it to the https://community.theforeman.org, where more people can get involved and express their opinions on this matter.

**Pull request merged**